### DISCUSSION TITLE
ESP32 - Cannot find reference 'SoftSPI' in 'machine.pyi'

### DISCUSSION THREAD

**Discussion by CobyC** [micropython/micropython#10107] [Libraries & Drivers]
Hi I'm trying out MicorPython on ESP32:
flashed with command
`esptool --chip esp32 --port COM3 write_flash -z 0x1000 H:\downloads\esp32-20220618-v1.19.1.bin`
terminal output:
esptool.py v3.0
Serial port COM3
Connecting........_
Chip is ESP32-D0WDQ6 (revision 1)
Features: WiFi, BT, Dual Core, 240MHz, VRef calibration in efuse, Coding Scheme None
Crystal is 40MHz

I'm using PyCharm 2022.2.4 and python 3.6 as interpreter. 

But when I try the example as shown in [quick references](https://docs.micropython.org/en/latest/esp32/quickref.html#software-spi-bus) 

on the line `from machine import Pin, SoftSPI`
I get a red line with a reference error Cannot find reference 'SoftSPI' in 'machine.pyi' 

Do I need to add another library?

**Comment by fivdi** [comment]
I can't reproduce the error on a ESP32-DevKitC-v4 when I connect directly to the ESP32 board with mpremote:
```
$ mpremote connect /dev/ttyUSB0 repl
Connected to MicroPython at /dev/ttyUSB0
Use Ctrl-] to exit this shell

>>> 
MicroPython v1.19.1-705-gac5934c96-dirty on 2022-11-18; ESP32 module with ESP32
Type "help()" for more information.
>>> from machine import Pin, SoftSPI
>>> help(SoftSPI)
object <class 'SoftSPI'> is of type type
  init -- <function>
  deinit -- <function>
  read -- <function>
  readinto -- <function>
  write -- <function>
  write_readinto -- <function>
  MSB -- 0
  LSB -- 1
>>> 
```
I'm not familiar with PyCharm, but given that you are using PyCharm, are you sure that you are actually running the code on the ESP32 board and not on your PC?

**Reply by CobyC** [reply]
Thanks for the reply, I think it is because I'm missing the stubs, so it looks like it's a pycharm issue.

**Comment by CobyC** [comment]
I think I found the issue. [PyCharm stubs](https://micropython-stubs.readthedocs.io/en/main/24_pycharm.html)

**Reply by Josverl** [reply]
Happy to see that you found the stubs and the docs :-) 
if you find the micropython-stubs docs unclear in any way, feel free to improve them or create an issue.
